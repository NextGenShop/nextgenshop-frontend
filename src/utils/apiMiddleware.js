import axios from "axios";
import config from "../../config";

// exampleRequest {
//   headers?: any;
//   data?: any;
//   url: string;
//   method: 'GET' | 'POST' | 'PUT' | 'DELETE' | 'PATCH';
// }

const apiMiddleware = ({ dispatch }) => (next) => (action) => {
  next(action);

  if (action.type !== "API") return;

  const { name, url, requestData, extraData } = action;
  axios.defaults.baseURL = config.API_URL;
  axios.defaults.headers.common["Content-Type"] = "application/json";

  let req = {
    url,
    method: requestData.method,
  };
  if (requestData.headers) {
    req.headers = requestData.headers;
  }
  if (requestData.data) {
    req.data = requestData.data;
  }

  dispatch({ type: `${name}_REQUEST` });
  axios
    .request(req)
    .then(({ data }) => {
      dispatch({ type: `${name}_SUCCESS`, response: data, extraData });
    })
    .catch((err) => {
      const errorMessage = err.response ? err.response.data.error : "An unknown error has occurred";
      dispatch({ type: `${name}_FAILURE`, error: errorMessage });
    });
};

export default apiMiddleware;
