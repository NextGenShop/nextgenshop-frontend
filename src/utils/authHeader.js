export default function authHeader(state) {
  const token = state.auth.token;
  if (token) {
    return { authorization: `Token ${token}` };
  } else {
    return null;
  }
}
