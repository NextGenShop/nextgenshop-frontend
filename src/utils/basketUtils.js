const calcTotalPrice = (items) => {
  let totalPrice = 0;
  for (let item of items) {
    totalPrice += item.product.price * item.quantity;
  }
  return totalPrice;
};

const addBasketItem = (product, basket, quantity = 1) => {
  let newItems = [];
  const foundItem = basket.items.find(
    (item) => item.product.productId === product.productId
  );
  if (foundItem) {
    newItems = basket.items.map((item) =>
      item.product.productId === foundItem.product.productId
        ? { ...item, quantity: item.quantity + quantity }
        : item
    );
  } else {
    newItems = [...basket.items, { product, quantity }];
  }
  return { items: newItems };
};

const removeBasketItem = (productId, basket) => {
  let newItems = basket.items.map((item) => {
    if (item.product.productId === productId) {
      if (item.quantity <= 1) {
        return null;
      } else {
        return { ...item, quantity: item.quantity - 1 };
      }
    }
    return item;
  });
  newItems = newItems.filter((item) => item !== null);
  return { items: newItems };
};

export { addBasketItem, removeBasketItem, calcTotalPrice };
