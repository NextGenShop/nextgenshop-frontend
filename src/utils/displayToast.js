import { toastr } from "react-redux-toastr";

let options = {
  timeOut: 5000,
  showCloseButton: true,
  progressBar: false,
  transitionIn: "fadeIn",
  transitionOut: "fadeOut",
};

export const displayToast = (title, msg, type, callbackClear) => {
  const toastrInstance =
    type === "info"
      ? toastr.info
      : type === "warning"
      ? toastr.warning
      : type === "danger"
      ? toastr.error
      : toastr.success;
  if (callbackClear) options.onShowComplete = callbackClear;
  toastrInstance(title, msg, options);
};
