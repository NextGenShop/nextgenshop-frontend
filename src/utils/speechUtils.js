import recognizeMic from "watson-speech/speech-to-text/recognize-microphone";
import synthesize from "watson-speech/text-to-speech/synthesize";

//audio/wav is supported in Chrome, Edge, Firefox (v3.5), Opera (v11.00), Safari (v3.1)
//audio/mp3 is supported in Chrome, IE9, Edge, Firefox (v71), Opera, Safari (v3.1)
//ref: https://en.wikipedia.org/wiki/HTML5_audio

export const getSupportedAudioFormat = (audioElement) => {
  if (!audioElement) {
    audioElement = document.createElement("audio");
  }

  let accept =
    audioElement &&
    audioElement.canPlayType === "function" &&
    audioElement.canPlayType("audio/mp3") !== ""
      ? "audio/mp3"
      : "audio/wav";

  return accept;
};

export const captureAudioFromMicrophone = async (
  speechToTextUrl,
  speechToTextToken
) => {
  let mediaStream = null;
  try {
    mediaStream = await navigator.mediaDevices.getUserMedia({
      video: false,
      audio: true,
    });
  } catch (err) {
    console.log(err);
  }

  const recognizeMicrophoneStream = recognizeMic({
    url: speechToTextUrl,
    accessToken: speechToTextToken,
    objectMode: true,
    extractResults: true,
    format: true,
    mediaStream,
    keepMicrophone: true,
  });

  return recognizeMicrophoneStream;
};

export const textToSpeech = async (
  textToSpeechUrl,
  textToSpeechToken,
  text,
  voice,
  accept,
  element
) => {
  return synthesize({
    url: textToSpeechUrl,
    accessToken: textToSpeechToken,
    text,
    voice,
    accept,
    element,
  });
};

export const createAssistantSession = async (assistantUrl, assistantToken) => {
  let session;
  try {
    const res = await fetch(`${assistantUrl}/sessions?version=2020-04-01`, {
      method: "post",
      headers: new Headers({
        Authorization: "Bearer " + assistantToken,
        "Content-Type": "application/json",
      }),
    });
    session = await res.json();
  } catch (err) {
    console.log(err);
  }
  return session ? session.session_id : null;
};

export const messageAssistant = async (
  assistantUrl,
  assistantToken,
  sessionId,
  message
) => {
  let res;
  let data = { speech: "", action: null };
  try {
    res = await fetch(
      `${assistantUrl}/sessions/${sessionId}/message?version=2020-04-01`,
      {
        method: "post",
        headers: new Headers({
          Authorization: "Bearer " + assistantToken,
          Accept: "application/json",
          "Content-Type": "application/json",
        }),
        body: JSON.stringify({ input: { text: message } }),
      }
    );
    res = await res.json();
    console.log(res);
    data.speech =
      '<speak version="1.0"><prosody rate="+15%">' +
      res.output.generic
        .reduce((acc, obj) => {
          if (obj.response_type === "text") {
            acc.push(obj.text);
          }
          return acc;
        }, [])
        .join(' <break strength="medium"></break>') +
      "</prosody></speak>";
    const actions = res.output.generic.reduce((acc, obj) => {
      if (obj.response_type === "user_defined") {
        if (obj.user_defined.quantity) {
          obj.user_defined.action_type = "add_basket";
        } else {
          obj.user_defined.action_type = "filter_product";
        }
        acc.push(obj.user_defined);
      }
      return acc;
    }, []);
    data.action = actions[0];
  } catch (err) {
    console.log(err);
  }
  console.log(data);
  return data;
};

export const stripSSMLTags = (str) => {
  const regex = /(<([^>]+)>)/gi;
  try {
    str = str.replace(regex, "");
  } catch (err) {
    console.log(err);
  }
  return str;
};
