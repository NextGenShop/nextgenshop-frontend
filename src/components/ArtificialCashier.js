import React from "react";
import { makeStyles } from "@material-ui/core/styles";
import Paper from "@material-ui/core/Paper";
import StopIcon from "@material-ui/icons/StopOutlined";
import IconButton from "@material-ui/core/IconButton";
import MicRoundedIcon from "@material-ui/icons/MicRounded";
import CircularProgress from "@material-ui/core/CircularProgress";
import Typography from "@material-ui/core/Typography";
import { connect } from "react-redux";
import { actions } from "../store/api/tokens";
import { actions as basketActions } from "../store/api/basket";
import { actions as productActions } from "../store/api/product";
import {
  getSupportedAudioFormat,
  captureAudioFromMicrophone,
  textToSpeech,
  createAssistantSession,
  messageAssistant,
  stripSSMLTags,
} from "../utils/speechUtils";
import { addBasketItem } from "../utils/basketUtils";
import AvatarModel from "../assets/models/Avatar.glb";
import { displayToast } from "../utils/displayToast";

const useStyles = makeStyles((theme) => ({
  paper: {
    display: "flex",
    color: theme.palette.text.secondary,
    background: "#eeeeee",
    height: "75vh",
    alignItems: "center",
    justifyContent: "center",
  },
  inner: {
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
  },
  bold: {
    fontWeight: "bold",
  },
  transcript: {
    wordWrap: "break-word",
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(2),
  },
}));

function ArtificialCashier({
  authUser,
  basket,
  products,
  speechToTextToken,
  speechToTextUrl,
  textToSpeechToken,
  textToSpeechUrl,
  assistantToken,
  assistantUrl,
  dispatchGetSpeechToTextToken,
  dispatchGetTextToSpeechToken,
  dispatchGetAssistantToken,
  dispatchUpdateBasket,
  dispatchGetProducts,
}) {
  const classes = useStyles();
  const [loaded, setLoaded] = React.useState(false);
  const [listening, setListening] = React.useState(false);
  const [responding, setResponding] = React.useState(false);
  const [stream, setStream] = React.useState(null);
  const [speechText, setSpeechText] = React.useState("");
  const [responseText, setResponseText] = React.useState("");
  const [assistantSessionId, setAssistantSessionId] = React.useState(null);
  const audioRef = React.useRef(null);

  const testVoices = [
    "en-GB_JamesV3Voice",
    "en-GB_KateV3Voice",
    "en-GB_CharlotteV3Voice",
  ];
  //ref: https://cloud.ibm.com/docs/text-to-speech?topic=text-to-speech-voices

  React.useEffect(() => {
    if (!speechToTextToken || !speechToTextUrl) {
      dispatchGetSpeechToTextToken();
    }
    if (!textToSpeechToken || !textToSpeechUrl) {
      dispatchGetTextToSpeechToken();
    }
    if (!assistantToken || !assistantUrl) {
      dispatchGetAssistantToken();
    }
  }, [
    dispatchGetSpeechToTextToken,
    dispatchGetTextToSpeechToken,
    dispatchGetAssistantToken,
    speechToTextToken,
    speechToTextUrl,
    textToSpeechToken,
    textToSpeechUrl,
    assistantToken,
    assistantUrl,
  ]);

  React.useEffect(() => {
    async function createSession() {
      const instanceId = await createAssistantSession(
        assistantUrl,
        assistantToken
      );
      setAssistantSessionId(instanceId);
      setLoaded(true);
      console.log("Done loading tokens from API.");
      console.log("Created assistant session: " + instanceId);
    }

    if (!!assistantToken && !!assistantUrl && !assistantSessionId) {
      createSession();
    }
  }, [assistantToken, assistantUrl, assistantSessionId]);

  const startListening = async () => {
    stopAudio();
    setListening(true);
    const stream = await captureAudioFromMicrophone(
      speechToTextUrl,
      speechToTextToken
    );
    stream.on("data", async (data) => {
      if (data && data.final) {
        const speechText = data.alternatives[0].transcript;
        setSpeechText(speechText);
        setListening(false);

        const res = await messageAssistant(
          assistantUrl,
          assistantToken,
          assistantSessionId,
          speechText
        );

        if (res.action) {
          handleAssistantAction(res.action);
        }

        if (res.speech) {
          await synthesizeTextToSpeech(res.speech);
        }
      }
    });
    stream.on("error", function (err) {
      console.log(err);
      setListening(false);
    });
    stream.on("end", function () {
      console.log("stream ended");
      setListening(false);
    });
    setStream(stream);
  };

  React.useEffect(() => {
    if (!listening && stream !== null) {
      stream.stop();
      setStream(null);
    }
  }, [listening, stream]);

  const handleAssistantAction = (action) => {
    switch (action.action_type) {
      case "add_basket":
        const product = products.find(
          (p) => p.productId === parseInt(action.product_id)
        );
        if (product) {
          const newBasket = addBasketItem(product, basket, action.quantity);
          dispatchUpdateBasket(authUser.userId, newBasket);
          displayToast("", "Your shopping basket has been updated", "success");
        }
        break;
      case "filter_product":
        if (action.product_query_string) {
          dispatchGetProducts(action.product_query_string, "Mock Retailer", 3);
          displayToast(
            "",
            "The product catalogue has been updated to match your query",
            "info"
          );
        }
        break;
      case "reset_context":
        resetProductCatalog();
        break;
      default:
        break;
    }
  };

  const resetProductCatalog = () => {
    dispatchGetProducts(null, "Mock Retailer", 9);
  };

  const handleAudioPlayingEvent = () => {
    console.log("AUDIO PLAYING!");
    setResponding(true);
  };

  const handleAudioEndedEvent = () => {
    console.log("AUDIO STOPPED PLAYING!");
    setResponding(false);
  };

  const synthesizeTextToSpeech = async (text, voice) => {
    if (!voice) voice = testVoices[0];
    try {
      const accept = getSupportedAudioFormat(audioRef);
      const audio = await textToSpeech(
        textToSpeechUrl,
        textToSpeechToken,
        text,
        voice,
        accept,
        audioRef.current
      );
      setResponseText(stripSSMLTags(text));
      audio.removeEventListener("playing", handleAudioPlayingEvent);
      audio.removeEventListener("ended", handleAudioEndedEvent);
      audio.addEventListener("playing", handleAudioPlayingEvent);
      audio.addEventListener("ended", handleAudioEndedEvent);
    } catch (err) {
      console.log(err);
    }
  };

  const stopAudio = () => {
    setResponseText("");
    audioRef.current.pause();
    audioRef.current.currentTime = 0;
  };

  const ListenButton = listening ? (
    <IconButton
      color="secondary"
      aria-label="stop"
      onClick={() => setListening(false)}
    >
      <StopIcon fontSize="large" />
    </IconButton>
  ) : (
    <IconButton color="primary" aria-label="speak" onClick={startListening}>
      <MicRoundedIcon fontSize="large" />
    </IconButton>
  );

  return (
    <Paper className={classes.paper}>
      <audio ref={audioRef}>
        Your browser does not support the <code>audio</code> element.
      </audio>
      <div className={classes.inner}>
        {loaded ? (
          <React.Fragment>
            {/* {speechText && (
              <Typography variant="h6">
                <span className={classes.bold}>Speech:</span> {speechText}
              </Typography>
            )} */}
            <model-viewer
              autoplay
              loading="eager"
              animation-name={responding ? "Talking" : "Idle"}
              style={{ width: "350px", height: "350px" }}
              src={AvatarModel}
            ></model-viewer>
            {responseText && (
              <Typography className={classes.transcript} variant="h6">
                {responseText}
              </Typography>
            )}
            <br />
            <Typography variant="subtitle2">
              {listening
                ? "Listening..."
                : "Press the mic icon and start speaking"}
            </Typography>
            {ListenButton}
          </React.Fragment>
        ) : (
          <React.Fragment>
            <Typography variant="subtitle2" gutterBottom>
              Loading... Please wait
            </Typography>
            <CircularProgress size={30} />
          </React.Fragment>
        )}
      </div>
    </Paper>
  );
}

const mapStateToProps = (state) => ({
  speechToTextToken: state.tokens.speechToTextToken,
  speechToTextUrl: state.tokens.speechToTextUrl,
  textToSpeechToken: state.tokens.textToSpeechToken,
  textToSpeechUrl: state.tokens.textToSpeechUrl,
  assistantToken: state.tokens.assistantToken,
  assistantUrl: state.tokens.assistantUrl,
  basket: state.basket,
  authUser: state.auth.user,
  products: state.product.products,
});

const mapDispatchToProps = (dispatch) => ({
  dispatchGetSpeechToTextToken: () => dispatch(actions.getSpeechToTextToken()),
  dispatchGetTextToSpeechToken: () => dispatch(actions.getTextToSpeechToken()),
  dispatchGetAssistantToken: () => dispatch(actions.getAssistantToken()),
  dispatchUpdateBasket: (shopperId, basketData) =>
    dispatch(basketActions.updateBasket(shopperId, basketData)),
  dispatchGetProducts: (query, retailer, limit) =>
    dispatch(productActions.getProducts(query, retailer, limit)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ArtificialCashier);
