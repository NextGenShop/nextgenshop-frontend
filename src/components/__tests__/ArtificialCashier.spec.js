import React from 'react';
import { shallow } from 'enzyme';
import { ArtificialCashier, mapDispatchToProps } from '../ArtificialCashier';
import mockBasketItems from '../../store/mock/MockBasketItems.json';
import mockProducts from '../../store/mock/MockSupermarketDataset.json';
import mockUsers from '../../store/mock/MockUsers.json';
import * as speechUtils from '../../utils/speechUtils';
import { displayToast } from '../../utils/displayToast';

jest.mock('../../store/api/tokens', () => {
  return {
    ...jest.requireActual('../../store/api/tokens'),
    actions: {
      getSpeechToTextToken: () => 'GET_SPEECH_TO_TEXT_TOKEN',
      getTextToSpeechToken: () => 'GET_TEXT_TO_SPEECH_TOKEN',
      getAssistantToken: () => 'GET_ASSISTANT_TOKEN',
    },
  };
});

jest.mock('../../store/api/basket', () => {
  return {
    ...jest.requireActual('../../store/api/basket'),
    actions: {
      updateBasket: () => 'UPDATE_BASKET',
    },
  };
});

jest.mock('../../store/api/product', () => {
  return {
    ...jest.requireActual('../../store/api/product'),
    actions: {
      getProducts: () => 'GET_PRODUCTS',
    },
  };
});

jest.mock('../../utils/displayToast', () => {
  return {
    displayToast: jest.fn(),
  };
});

describe('ArtificialCashier', () => {
  const defaultProps = {
    authUser: mockUsers[0],
    basket: { items: mockBasketItems, totalPrice: 0, shopper: 0 },
    products: mockProducts,
    speechToTextToken: '',
    speechToTextUrl: '',
    textToSpeechToken: '',
    textToSpeechUrl: '',
    assistantToken: '',
    assistantUrl: '',
    dispatchGetSpeechToTextToken: jest.fn(),
    dispatchGetTextToSpeechToken: jest.fn(),
    dispatchGetAssistantToken: jest.fn(),
    dispatchUpdateBasket: jest.fn(),
    dispatchGetProducts: jest.fn(),
    history: { push: jest.fn() },
  };

  const loadedProps = {
    ...defaultProps,
    speechToTextToken: 'mockSpeechToTextToken',
    speechToTextUrl: 'mockSpeechToTextUrl',
    textToSpeechToken: 'mockTextToSpeechToken',
    textToSpeechUrl: 'mockTextToSpeechUrl',
    assistantToken: 'mockAssistantToken',
    assistantUrl: 'mockAssistantUrl',
  };

  const flushPromises = () => new Promise(setImmediate);
  const realUseState = React.useState;
  const reactMock = require('react');

  const mockUseState = (loaded = false) => {
    reactMock.useState = jest
      .fn()
      .mockImplementationOnce(() => realUseState(loaded)) //loaded
      .mockImplementationOnce(() => realUseState(false)) //listening
      .mockImplementationOnce(() => realUseState(false)) //responding
      .mockImplementationOnce(() => realUseState(null)) //stream
      .mockImplementationOnce(() => realUseState('')) //speechText
      .mockImplementationOnce(() => realUseState('')) //responseText
      .mockImplementationOnce(() => realUseState(loaded ? 'mockSessionId' : '')) //assistantSessionId
      .mockImplementationOnce(() => realUseState(false)) //redirectCheckout
      .mockImplementationOnce(() => realUseState(0)); //voiceIndex
  };

  beforeEach(() => {
    jest.spyOn(React, 'useEffect').mockImplementation((f) => f());
    jest
      .spyOn(speechUtils, 'getSupportedAudioFormat')
      .mockImplementation(() => 'audio/wav');
    jest
      .spyOn(speechUtils, 'captureAudioFromMicrophone')
      .mockImplementation(() =>
        Promise.resolve({
          on: (str, fn) => {
            const mockData = {
              final: true,
              alternatives: [{ transcript: 'mockSpeechText' }],
            };
            if (str === 'data') fn(mockData);
          },
          stop: () => {},
        })
      );
    jest.spyOn(speechUtils, 'textToSpeech').mockImplementation(() =>
      Promise.resolve({
        addEventListener: () => {},
        removeEventListener: () => {},
      })
    );
    jest
      .spyOn(speechUtils, 'createAssistantSession')
      .mockImplementation(() => 'mockSessionId');
    jest
      .spyOn(speechUtils, 'messageAssistant')
      .mockImplementation(() => Promise.resolve({ speech: '', actions: [] }));
    jest.spyOn(speechUtils, 'stripSSMLTags').mockImplementation((str) => str);
  });

  test('should correctly render with props', () => {
    const component = shallow(<ArtificialCashier {...defaultProps} />);
    expect(component).toMatchSnapshot();
  });

  test('should correctly fetch tokens', () => {
    shallow(<ArtificialCashier {...defaultProps} />);

    expect(defaultProps.dispatchGetSpeechToTextToken).toBeCalledTimes(1);
    expect(defaultProps.dispatchGetTextToSpeechToken).toBeCalledTimes(1);
    expect(defaultProps.dispatchGetAssistantToken).toBeCalledTimes(1);
  });

  test('should render avatar model when loaded', () => {
    mockUseState(true);
    const component = shallow(<ArtificialCashier {...defaultProps} />);
    expect(component.exists('model-viewer')).toEqual(true);
    expect(component.exists('#loadingText')).toEqual(false);
  });

  test('should render loading text when loading', () => {
    mockUseState(false);
    const component = shallow(<ArtificialCashier {...defaultProps} />);
    expect(component.exists('model-viewer')).toEqual(false);
    expect(component.exists('#loadingText')).toEqual(true);
  });

  test('should correctly handle start listening', async () => {
    mockUseState(true);
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.find('#listenText').first().text()).toEqual(
      'Press the mic icon and start speaking'
    );
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState; // reset mocked useState to realUseState
    component.find('#startListenButton').first().simulate('click');
    expect(component.find('#listenText').first().text()).toEqual(
      'Listening...'
    );
    await flushPromises(); //waits for promises to resolve
    component.update(); //forces component to re-render
    expect(speechUtils.captureAudioFromMicrophone).toBeCalledWith(
      'mockSpeechToTextUrl',
      'mockSpeechToTextToken'
    );
    expect(speechUtils.messageAssistant).toBeCalledWith(
      'mockAssistantUrl',
      'mockAssistantToken',
      'mockSessionId',
      'mockSpeechText'
    );
    expect(speechUtils.textToSpeech).not.toBeCalled();
  });

  test('should correctly handle stop listening', () => {
    mockUseState(true);
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.find('#listenText').first().text()).toEqual(
      'Press the mic icon and start speaking'
    );
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState;
    component.find('#startListenButton').first().simulate('click');
    expect(component.find('#listenText').first().text()).toEqual(
      'Listening...'
    );
    expect(component.exists('#stopListenButton')).toEqual(true);
    component.find('#stopListenButton').first().simulate('click');
    expect(component.find('#listenText').first().text()).toEqual(
      'Press the mic icon and start speaking'
    );
  });

  test('should correctly call text to speech', async () => {
    mockUseState(true);
    jest
      .spyOn(speechUtils, 'messageAssistant')
      .mockImplementation(() =>
        Promise.resolve({ speech: 'mockSpeechText', actions: [] })
      );
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState;
    component.find('#startListenButton').first().simulate('click');
    await flushPromises();
    component.update();
    expect(speechUtils.textToSpeech).toBeCalled();
  });

  test('should correctly handle add_basket action', async () => {
    mockUseState(true);
    jest.spyOn(speechUtils, 'messageAssistant').mockImplementation(() =>
      Promise.resolve({
        speech: '',
        actions: [
          {
            action_type: 'add_basket',
            product_id: mockProducts[0].productId,
            quantity: 1,
          },
        ],
      })
    );
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState;
    component.find('#startListenButton').first().simulate('click');
    await flushPromises();
    component.update();
    expect(defaultProps.dispatchUpdateBasket).toBeCalled();
    expect(displayToast).toBeCalled();
  });

  test('should correctly handle filter_product action', async () => {
    mockUseState(true);
    jest.spyOn(speechUtils, 'messageAssistant').mockImplementation(() =>
      Promise.resolve({
        speech: '',
        actions: [
          {
            action_type: 'filter_product',
            product_query_string: 'mockQueryString',
          },
        ],
      })
    );
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState;
    component.find('#startListenButton').first().simulate('click');
    await flushPromises();
    component.update();
    expect(defaultProps.dispatchGetProducts).toBeCalled();
    expect(displayToast).toBeCalled();
  });

  test('should correctly handle reset_context action', async () => {
    mockUseState(true);
    jest.spyOn(speechUtils, 'messageAssistant').mockImplementation(() =>
      Promise.resolve({
        speech: '',
        actions: [
          {
            action_type: 'reset_context',
          },
        ],
      })
    );
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState;
    component.find('#startListenButton').first().simulate('click');
    await flushPromises();
    component.update();
    expect(defaultProps.dispatchGetProducts).toBeCalledWith(
      null,
      'Mock Retailer',
      9
    );
  });

  test('should correctly handle checkout action', async () => {
    mockUseState(true);
    jest.spyOn(speechUtils, 'messageAssistant').mockImplementation(() =>
      Promise.resolve({
        speech: '',
        actions: [
          {
            action_type: 'checkout',
          },
        ],
      })
    );
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState;
    component.find('#startListenButton').first().simulate('click');
    await flushPromises();
    component.update();
    expect(defaultProps.history.push).toBeCalledWith({
      pathname: '/checkout',
    });
  });

  test('should ignore invalid actions', async () => {
    mockUseState(true);
    jest.spyOn(speechUtils, 'messageAssistant').mockImplementation(() =>
      Promise.resolve({
        speech: '',
        actions: [
          {
            action_type: 'invalidActionType',
          },
          {
            action_type: 'invalidActionTypeB',
          },
        ],
      })
    );
    const component = shallow(<ArtificialCashier {...loadedProps} />);
    expect(component.exists('#startListenButton')).toEqual(true);
    reactMock.useState = realUseState;
    component.find('#startListenButton').first().simulate('click');
    await flushPromises();
    component.update();
    expect(defaultProps.dispatchGetProducts).not.toBeCalled();
  });

  test('should dispatch correct actions', () => {
    const dispatch = jest.fn();
    mapDispatchToProps(dispatch).dispatchGetSpeechToTextToken();
    mapDispatchToProps(dispatch).dispatchGetTextToSpeechToken();
    mapDispatchToProps(dispatch).dispatchGetAssistantToken();
    mapDispatchToProps(dispatch).dispatchUpdateBasket();
    mapDispatchToProps(dispatch).dispatchGetProducts();
    expect(dispatch.mock.calls[0][0]).toEqual('GET_SPEECH_TO_TEXT_TOKEN');
    expect(dispatch.mock.calls[1][0]).toBe('GET_TEXT_TO_SPEECH_TOKEN');
    expect(dispatch.mock.calls[2][0]).toBe('GET_ASSISTANT_TOKEN');
    expect(dispatch.mock.calls[3][0]).toBe('UPDATE_BASKET');
    expect(dispatch.mock.calls[4][0]).toBe('GET_PRODUCTS');
  });
});
