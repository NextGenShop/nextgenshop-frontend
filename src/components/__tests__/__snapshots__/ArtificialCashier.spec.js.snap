// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArtificialCashier should correctly render with props 1`] = `
<WithStyles(ForwardRef(Paper))
  className="makeStyles-paper-1"
>
  <audio>
    Your browser does not support the 
    <code>
      audio
    </code>
     element.
  </audio>
  <div
    className="makeStyles-inner-2"
  >
    <WithStyles(ForwardRef(Typography))
      gutterBottom={true}
      id="loadingText"
      variant="subtitle2"
    >
      Loading... Please wait
    </WithStyles(ForwardRef(Typography))>
    <WithStyles(ForwardRef(CircularProgress))
      size={30}
    />
  </div>
</WithStyles(ForwardRef(Paper))>
`;
